// @microsoft/kiota/dist/esm/types.js
export var KiotaGenerationLanguage;
(function (KiotaGenerationLanguage) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["CSharp"] = 0] = "CSharp";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["Java"] = 1] = "Java";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["TypeScript"] = 2] = "TypeScript";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["PHP"] = 3] = "PHP";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["Python"] = 4] = "Python";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["Go"] = 5] = "Go";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["Swift"] = 6] = "Swift";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["Ruby"] = 7] = "Ruby";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["CLI"] = 8] = "CLI";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaGenerationLanguage[KiotaGenerationLanguage["Dart"] = 9] = "Dart";
})(KiotaGenerationLanguage || (KiotaGenerationLanguage = {}));
;
export var KiotaPluginType;
(function (KiotaPluginType) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaPluginType[KiotaPluginType["OpenAI"] = 0] = "OpenAI";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaPluginType[KiotaPluginType["ApiManifest"] = 1] = "ApiManifest";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    KiotaPluginType[KiotaPluginType["ApiPlugin"] = 2] = "ApiPlugin";
})(KiotaPluginType || (KiotaPluginType = {}));
;
export var OpenApiAuthType;
(function (OpenApiAuthType) {
    OpenApiAuthType[OpenApiAuthType["None"] = 0] = "None";
    OpenApiAuthType[OpenApiAuthType["ApiKey"] = 1] = "ApiKey";
    OpenApiAuthType[OpenApiAuthType["Http"] = 2] = "Http";
    OpenApiAuthType[OpenApiAuthType["OAuth2"] = 3] = "OAuth2";
    OpenApiAuthType[OpenApiAuthType["OpenIdConnect"] = 4] = "OpenIdConnect";
})(OpenApiAuthType || (OpenApiAuthType = {}));
export var OpenApiSpecVersion;
(function (OpenApiSpecVersion) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    OpenApiSpecVersion[OpenApiSpecVersion["V2_0"] = 0] = "V2_0";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    OpenApiSpecVersion[OpenApiSpecVersion["V3_0"] = 1] = "V3_0";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    OpenApiSpecVersion[OpenApiSpecVersion["V3_1"] = 2] = "V3_1";
})(OpenApiSpecVersion || (OpenApiSpecVersion = {}));
export var ConsumerOperation;
(function (ConsumerOperation) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    ConsumerOperation[ConsumerOperation["Add"] = 0] = "Add";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    ConsumerOperation[ConsumerOperation["Edit"] = 1] = "Edit";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    ConsumerOperation[ConsumerOperation["Remove"] = 2] = "Remove";
    // eslint-disable-next-line @typescript-eslint/naming-convention
    ConsumerOperation[ConsumerOperation["Generate"] = 3] = "Generate";
})(ConsumerOperation || (ConsumerOperation = {}));
export function generationLanguageToString(language) {
    switch (language) {
        case KiotaGenerationLanguage.CSharp:
            return "CSharp";
        case KiotaGenerationLanguage.Java:
            return "Java";
        case KiotaGenerationLanguage.TypeScript:
            return "TypeScript";
        case KiotaGenerationLanguage.PHP:
            return "PHP";
        case KiotaGenerationLanguage.Python:
            return "Python";
        case KiotaGenerationLanguage.Go:
            return "Go";
        case KiotaGenerationLanguage.Swift:
            return "Swift";
        case KiotaGenerationLanguage.Ruby:
            return "Ruby";
        case KiotaGenerationLanguage.CLI:
            return "CLI";
        case KiotaGenerationLanguage.Dart:
            return "Dart";
        default:
            throw new Error("unknown language");
    }
}
export const allGenerationLanguages = [
    KiotaGenerationLanguage.CSharp,
    KiotaGenerationLanguage.Go,
    KiotaGenerationLanguage.Java,
    KiotaGenerationLanguage.PHP,
    KiotaGenerationLanguage.Python,
    KiotaGenerationLanguage.Ruby,
    KiotaGenerationLanguage.CLI,
    KiotaGenerationLanguage.Swift,
    KiotaGenerationLanguage.TypeScript,
    KiotaGenerationLanguage.Dart,
];
/**
 * The log level from Kiota
 * @see https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loglevel?view=dotnet-plat-ext-7.0
 */
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["trace"] = 0] = "trace";
    LogLevel[LogLevel["debug"] = 1] = "debug";
    LogLevel[LogLevel["information"] = 2] = "information";
    LogLevel[LogLevel["warning"] = 3] = "warning";
    LogLevel[LogLevel["error"] = 4] = "error";
    LogLevel[LogLevel["critical"] = 5] = "critical";
    LogLevel[LogLevel["none"] = 6] = "none";
})(LogLevel || (LogLevel = {}));
export var MaturityLevel;
(function (MaturityLevel) {
    MaturityLevel[MaturityLevel["experimental"] = 0] = "experimental";
    MaturityLevel[MaturityLevel["preview"] = 1] = "preview";
    MaturityLevel[MaturityLevel["stable"] = 2] = "stable";
})(MaturityLevel || (MaturityLevel = {}));
export var DependencyType;
(function (DependencyType) {
    DependencyType[DependencyType["abstractions"] = 0] = "abstractions";
    DependencyType[DependencyType["serialization"] = 1] = "serialization";
    DependencyType[DependencyType["authentication"] = 2] = "authentication";
    DependencyType[DependencyType["http"] = 3] = "http";
    DependencyType[DependencyType["bundle"] = 4] = "bundle";
    DependencyType[DependencyType["additional"] = 5] = "additional";
})(DependencyType || (DependencyType = {}));
export var PluginAuthType;
(function (PluginAuthType) {
    PluginAuthType["oAuthPluginVault"] = "OAuthPluginVault";
    PluginAuthType["apiKeyPluginVault"] = "ApiKeyPluginVault";
})(PluginAuthType || (PluginAuthType = {}));
